js/ui/windowManager.js

```diff
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
+            session.buttonApp.scale_y = 1;
+            session.buttonApp.scale_x = 1;
+            session.buttonApp.translation_y = 0;
+            session.buttonApp.translation_x = 0;
+            this._rotateInActors.push(session.windowBuilder);
+            this._rotateOutActors.push(session.windowApp);
+            this._animate(session.windowApp, session.windowBuilder, Gtk.DirectionType.LEFT);
+            this._firstFrameConnections = this._firstFrameConnections.filter(function(conn) {
+                return conn != session.windowBuilder._firstFrameConnection;
+            });
+            session.windowBuilder.disconnect(session.windowBuilder._firstFrameConnection);
+            session.windowBuilder._firstFrameConnection = null;
+            this._addButton(session);
+            session.buttonBuilder.show();
+            session.buttonApp.hide();
+            return false;
+        }));
+        /* Save the connection's id on the destination window and in a list too so we can
+         * get rid of it on kill-window-effects later */
+        session.windowBuilder._firstFrameConnection = firstFrameConnection;
+        this._firstFrameConnections.push(firstFrameConnection);
+    },
+
+    _windowAppPositionChanged: function(window, event, session) {
+        let rect = session.windowApp.get_meta_window().get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowBuilder)
+                return;
+        session.windowBuilder.get_meta_window().move_resize_frame(false,
+                                                                  rect.x,
+                                                                  rect.y,
+                                                                  rect.width,
+                                                                  rect.height);
+    },
+
+    _windowBuilderPositionChanged: function(window, event, session) {
+        let rect = session.windowBuilder.get_meta_window().get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowApp)
+            return;
+        session.windowApp.get_meta_window().move_resize_frame(false,
+                                                              rect.x,
+                                                              rect.y,
+                                                              rect.width,
+                                                              rect.height);
+    },
+
+    _windowAppRestacked : function(overview, stackIndices, session) {
+        let focusedWindow = global.display.get_focus_window();
+        if (!focusedWindow)
+            return;
+        if (!session.windowApp.get_meta_window())
+            return;
+
+        if (session.buttonApp)
+            session.buttonApp.hide();
+        if (session.buttonBuilder)
```

Comment:
Also -> Are we destroying the buttons as soon as the window flips around? Or are we just keeping them around hidden? There's lots of null-checks here which may not be necessary (and otherwise confuse the reader).